<!DOCTYPE html>
<html lang="en"><head>
<script src="mariner_files/libs/clipboard/clipboard.min.js"></script>
<script src="mariner_files/libs/quarto-html/tabby.min.js"></script>
<script src="mariner_files/libs/quarto-html/popper.min.js"></script>
<script src="mariner_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="mariner_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="mariner_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="mariner_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.0.37">

  <title>mariner</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="mariner_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="mariner_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
  <link rel="stylesheet" href="mariner_files/libs/revealjs/dist/theme/quarto.css" id="theme">
  <link href="mariner_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="mariner_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="mariner_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="mariner_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="mariner_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="mariner_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-captioned.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-captioned) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-captioned.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-captioned .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-captioned .callout-caption  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-captioned.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-captioned.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-caption {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-caption {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-caption {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-captioned .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-captioned .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-captioned) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-caption {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-caption {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-caption {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-caption {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-caption {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">


<section id="section" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title"></h2>
<div class="header">
<img src="images/mariner_no_text_cropped.png">
<h1>
Mariner
</h1>
<div>

<div class="subheader absolute" style="top: 65%; left: 40%; ">
<p>BCB Student Seminar</p>
<p>Eric Davis</p>
<p>Phanstiel Lab</p>
<p>09/01/2022</p>
</div>
</div></div></section>
<section id="section-1" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title"></h2>
<div class="header">
<img src="images/mariner_no_text_cropped.png">
<h1 style="top:0px;margin-bottom:0px;color:#ff9865;">
Mariner
</h1>
<h4>
“Explore the Hi-C’s”
</h4>
<div>

<div class="subheader absolute" style="top: 65%; left: 40%; ">
<p>BCB Student Seminar</p>
<p>Eric Davis</p>
<p>Phanstiel Lab</p>
<p>09/01/2022</p>
</div>
<!-- Why do we need this? -->
</div></div></section>
<section>
<section id="why-do-we-need-this" class="title-slide slide level1 center middle">
<h1>Why do we need this?</h1>

</section>
<section id="why-do-we-need-this-1" class="slide level2">
<h2>Why do we need this?</h2>
<p><img data-src="https://imgs.xkcd.com/comics/standards.png" class="absolute" style="top: 15%; left: 15%; width: 70%; "></p>

<!-- Table Existing software -->
</section>
<section id="why-do-we-need-this-2" class="slide level2 smaller">
<h2>Why do we need this?</h2>

<table class="tg">
<thead>
<tr>
<th class="tg-nrix" rowspan="2">
Software
</th>
<th class="tg-nrix" rowspan="2">
Language
</th>
<th class="tg-nrix" colspan="3">
Extracting
</th>
<th class="tg-nrix" colspan="3">
Aggregating
</th>
<th class="tg-nrix" rowspan="2">
Range Operations
</th>
<th class="tg-nrix" rowspan="2">
Differential Analysis
</th>
<th class="tg-nrix" rowspan="2">
Visualization
</th>
</tr>
<tr>
<th class="tg-57iy">
Pixels
</th>
<th class="tg-57iy">
Matrices
</th>
<th class="tg-57iy">
Custom
</th>
<th class="tg-57iy">
Matrices
</th>
<th class="tg-57iy">
TADs/Stripes
</th>
<th class="tg-57iy">
Custom
</th>
</tr>
</thead>
<tbody>
<tr>
<td class="tg-nrix">
<a href="https://github.com/aidenlab/juicer">Juicer tools</a>
</td>
<td class="tg-nrix">
<img data-src="images/terminal.png" width="25"> <img data-src="images/javalogo.svg" width="25">
</td>
<td class="tg-nrix">
✅
</td>
<td class="tg-nrix">
✅
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
✅
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
❌
</td>
</tr>
<tr>
<td class="tg-57iy">
<a href="https://bmcbioinformatics.biomedcentral.com/track/pdf/10.1186/s12859-022-04589-y.pdf">HiCmapTools</a>
</td>
<td class="tg-57iy">
<img data-src="images/terminal.png" width="25">
</td>
<td class="tg-57iy">
✅
</td>
<td class="tg-57iy">
✅
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
✅
</td>
<td class="tg-57iy">
✅
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
⚠️
</td>
</tr>
<tr>
<td class="tg-nrix">
<a href="https://github.com/open2c/coolpuppy">coolpup.py</a>
</td>
<td class="tg-nrix">
<img data-src="images/pythonLogo.webp" width="25">
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
✅
</td>
<td class="tg-nrix">
✅
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
⚠️
</td>
</tr>
<tr>
<td class="tg-57iy">
<a href="https://aidenlab.org/juicebox/">Juice Box</a>
</td>
<td class="tg-57iy">
<img data-src="images/chrome.png" width="25">
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
✅
</td>
</tr>
<tr>
<td class="tg-nrix">
<a href="http://higlass.io/">HiGlass</a>
</td>
<td class="tg-nrix">
<img data-src="images/chrome.png" width="25">
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
❌
</td>
<td class="tg-nrix">
✅
</td>
</tr>
<tr>
<td class="tg-57iy">
<a href="https://bedtools.readthedocs.io/en/latest/">Bedtools</a>
</td>
<td class="tg-57iy">
<img data-src="images/terminal.png" width="25"> <img data-src="images/RLogo.webp" width="25">
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
✅
</td>
<td class="tg-57iy">
❌
</td>
<td class="tg-57iy">
❌
</td>
</tr>
</tbody>

</table>
<ul>
<li class="fragment"><p>Hi-C tools are <strong>dispersed</strong> across programming languages/environments.</p>
<ul>
<li class="fragment">Steep learning curve</li>
<li class="fragment">Difficult to create reproducible workflow.</li>
</ul></li>
<li class="fragment"><p>Some functionality is either impossible or impractical.</p></li>
</ul>
<!-- Bioconductor environment -->
</section></section>
<section>
<section id="advantages-of-the-bioconductor-environment" class="title-slide slide level1 center middle">
<h1>Advantages of the Bioconductor environment</h1>

</section>
<section id="bioconductor-environment" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">Bioconductor environment</h2>
<p><img data-src="images/bioclogo.png" class="absolute" style="top: 0px; right: 0px; width: 150px; "></p>
<ul>
<li class="fragment">Keep all analyses in R</li>
<li class="fragment">Classes for working with genomic data</li>
</ul>
<div class="fragment">
<div class="sourceCode" id="cb1" data-code-line-numbers="|3"><pre class="sourceCode numberSource r number-lines code-with-copy" data-id="quarto-animate-code"><code class="sourceCode r hljs"><span id="cb1-1" class="hljs-ln-code"><a href="#cb1-1"></a><span class="do">## Genomic ranges</span></span>
<span id="cb1-2" class="hljs-ln-code"><a href="#cb1-2"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb1-3" class="hljs-ln-code"><a href="#cb1-3"></a><span class="fu">GRanges</span>(<span class="st">"chr1:1-10000"</span>)</span>
<span id="cb1-4" class="hljs-ln-code"><a href="#cb1-4"></a></span>
<span id="cb1-5" class="hljs-ln-code"><a href="#cb1-5"></a><span class="do">## Alternate method</span></span>
<span id="cb1-6" class="hljs-ln-code"><a href="#cb1-6"></a><span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>),</span>
<span id="cb1-7" class="hljs-ln-code"><a href="#cb1-7"></a>        <span class="at">ranges =</span> <span class="fu">IRanges</span>(<span class="at">start =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">30000</span>),</span>
<span id="cb1-8" class="hljs-ln-code"><a href="#cb1-8"></a>                         <span class="at">end =</span> <span class="fu">c</span>(<span class="dv">10000</span>, <span class="dv">60000</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1   1-10000      *
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</div>
</section>
<section id="bioconductor-environment-1" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">Bioconductor environment</h2>
<p><img data-src="images/bioclogo.png" class="absolute" style="top: 0px; right: 0px; width: 150px; "></p>
<div>
<ul>
<li>Keep all analyses in R</li>
<li>Classes for working with genomic data</li>
</ul>
</div>
<div class="sourceCode" id="cb3" data-code-line-numbers="6-9"><pre class="sourceCode numberSource r number-lines code-with-copy" data-id="quarto-animate-code"><code class="sourceCode r hljs"><span id="cb3-1" class="hljs-ln-code"><a href="#cb3-1"></a><span class="do">## Genomic ranges</span></span>
<span id="cb3-2" class="hljs-ln-code"><a href="#cb3-2"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb3-3" class="hljs-ln-code"><a href="#cb3-3"></a><span class="fu">GRanges</span>(<span class="st">"chr1:1-10000"</span>)</span>
<span id="cb3-4" class="hljs-ln-code"><a href="#cb3-4"></a></span>
<span id="cb3-5" class="hljs-ln-code"><a href="#cb3-5"></a><span class="do">## Alternate method</span></span>
<span id="cb3-6" class="hljs-ln-code"><a href="#cb3-6"></a><span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>),</span>
<span id="cb3-7" class="hljs-ln-code"><a href="#cb3-7"></a>        <span class="at">ranges =</span> <span class="fu">IRanges</span>(<span class="at">start =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">30000</span>),</span>
<span id="cb3-8" class="hljs-ln-code"><a href="#cb3-8"></a>                         <span class="at">end =</span> <span class="fu">c</span>(<span class="dv">10000</span>, <span class="dv">60000</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="fragment">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 2 ranges and 0 metadata columns:
      seqnames      ranges strand
         &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1     1-10000      *
  [2]     chr2 30000-60000      *
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</div>
</section>
<section id="bioconductor-environment-2" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">Bioconductor environment</h2>
<p><img data-src="images/bioclogo.png" class="absolute" style="top: 0px; right: 0px; width: 150px; "></p>
<div>
<ul>
<li>Keep all analyses in R</li>
<li>Classes for working with genomic data</li>
</ul>
</div>
<div class="sourceCode" id="cb5" data-code-line-numbers="6-9|12-13"><pre class="sourceCode numberSource r number-lines code-with-copy" data-id="quarto-animate-code"><code class="sourceCode r hljs"><span id="cb5-1" class="hljs-ln-code"><a href="#cb5-1"></a><span class="do">## Genomic ranges</span></span>
<span id="cb5-2" class="hljs-ln-code"><a href="#cb5-2"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb5-3" class="hljs-ln-code"><a href="#cb5-3"></a><span class="fu">GRanges</span>(<span class="st">"chr1:1-10000"</span>)</span>
<span id="cb5-4" class="hljs-ln-code"><a href="#cb5-4"></a></span>
<span id="cb5-5" class="hljs-ln-code"><a href="#cb5-5"></a><span class="do">## Alternate method</span></span>
<span id="cb5-6" class="hljs-ln-code"><a href="#cb5-6"></a><span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>),</span>
<span id="cb5-7" class="hljs-ln-code"><a href="#cb5-7"></a>        <span class="at">ranges =</span> <span class="fu">IRanges</span>(<span class="at">start =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">30000</span>),</span>
<span id="cb5-8" class="hljs-ln-code"><a href="#cb5-8"></a>                         <span class="at">end =</span> <span class="fu">c</span>(<span class="dv">10000</span>, <span class="dv">60000</span>)))</span>
<span id="cb5-9" class="hljs-ln-code"><a href="#cb5-9"></a></span>
<span id="cb5-10" class="hljs-ln-code"><a href="#cb5-10"></a><span class="do">## Genomic interactions</span></span>
<span id="cb5-11" class="hljs-ln-code"><a href="#cb5-11"></a><span class="fu">library</span>(InteractionSet)</span>
<span id="cb5-12" class="hljs-ln-code"><a href="#cb5-12"></a><span class="fu">GInteractions</span>(<span class="at">anchor1 =</span> <span class="fu">GRanges</span>(<span class="st">"chr1:1-10000"</span>),</span>
<span id="cb5-13" class="hljs-ln-code"><a href="#cb5-13"></a>              <span class="at">anchor2 =</span> <span class="fu">GRanges</span>(<span class="st">"chr1:1-10000"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="fragment">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>GInteractions object with 1 interaction and 0 metadata columns:
      seqnames1   ranges1     seqnames2   ranges2
          &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt; &lt;IRanges&gt;
  [1]      chr1   1-10000 ---      chr1   1-10000
  -------
  regions: 1 ranges and 0 metadata columns
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</div>
</section></section>
<section>
<section id="mariner-functionality" class="title-slide slide level1 center middle">
<h1>Mariner functionality</h1>

</section>
<section id="manipulatingmerging-pairs" class="slide level2">
<h2>Manipulating/Merging Pairs</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="sourceCode" id="cb7" data-code-line-numbers="1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="fu">as_ginteractions</span>(df, ...)</span>
<span id="cb7-2"><a href="#cb7-2"></a></span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="fu">binPairs</span>(x, binSize, pos1, pos2, ...) </span>
<span id="cb7-4"><a href="#cb7-4"></a></span>
<span id="cb7-5"><a href="#cb7-5"></a><span class="fu">filterPairs</span>(x, binSize, ...)</span>
<span id="cb7-6"><a href="#cb7-6"></a></span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="fu">mergePairs</span>(x, radius, ...)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><br> <span style="font-size:22pt">Convert <code>data.frame</code> to <code>GInteractions</code></span></p>
</div><div class="column middle" style="width:50%;">
<p><img data-src="images/paste-B9258AD3.png" style="width:50.0%"></p>
</div>
</div>
</section>
<section id="manipulatingmerging-pairs-1" class="slide level2">
<h2>Manipulating/Merging Pairs</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="sourceCode" id="cb8" data-code-line-numbers="3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="fu">as_ginteractions</span>(df, ...)</span>
<span id="cb8-2"><a href="#cb8-2"></a></span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="fu">binPairs</span>(x, binSize, pos1, pos2, ...) </span>
<span id="cb8-4"><a href="#cb8-4"></a></span>
<span id="cb8-5"><a href="#cb8-5"></a><span class="fu">filterPairs</span>(x, binSize, ...)</span>
<span id="cb8-6"><a href="#cb8-6"></a></span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="fu">mergePairs</span>(x, radius, ...)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><br> <span style="font-size:22pt">Assign paired-ranges to specified <code>binSize</code></span></p>
</div><div class="column middle" style="width:50%;">
<p><img data-src="images/paste-B9258AD3.png" style="width:50.0%"></p>
</div>
</div>
</section>
<section id="manipulatingmerging-pairs-2" class="slide level2">
<h2>Manipulating/Merging Pairs</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="sourceCode" id="cb9" data-code-line-numbers="5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="fu">as_ginteractions</span>(df, ...)</span>
<span id="cb9-2"><a href="#cb9-2"></a></span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="fu">binPairs</span>(x, binSize, pos1, pos2, ...) </span>
<span id="cb9-4"><a href="#cb9-4"></a></span>
<span id="cb9-5"><a href="#cb9-5"></a><span class="fu">filterPairs</span>(x, binSize, buffer, ...)</span>
<span id="cb9-6"><a href="#cb9-6"></a></span>
<span id="cb9-7"><a href="#cb9-7"></a><span class="fu">mergePairs</span>(x, radius, ...)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><br> <span style="font-size:22pt">Filter out paired-ranges that would intersect the diagonal</span></p>
</div><div class="column middle" style="width:50%;">
<p><img data-src="images/filterPairs.png" style="width:50.0%"></p>
</div>
</div>
</section>
<section id="manipulatingmerging-pairs-3" class="slide level2">
<h2>Manipulating/Merging Pairs</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="sourceCode" id="cb10" data-code-line-numbers="7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="fu">as_ginteractions</span>(df, ...)</span>
<span id="cb10-2"><a href="#cb10-2"></a></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="fu">binPairs</span>(x, binSize, pos1, pos2, ...) </span>
<span id="cb10-4"><a href="#cb10-4"></a></span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="fu">filterPairs</span>(x, binSize, buffer, ...)</span>
<span id="cb10-6"><a href="#cb10-6"></a></span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="fu">mergePairs</span>(x, radius, ...)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><br> <span style="font-size:22pt">Cluster pairs with DBSCAN, then select a representative pair.</span></p>
</div><div class="column middle" style="width:50%;">
<p><img data-src="images/mergePairs.png" style="width:80.0%"></p>
</div>
</div>
</section>
<section id="generating-pairs-from-ranges" class="slide level2">
<h2>Generating Pairs from Ranges</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="sourceCode" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="fu">rangesToPairs</span>(x, window, ...)</span>
<span id="cb11-2"><a href="#cb11-2"></a></span>
<span id="cb11-3"><a href="#cb11-3"></a><span class="fu">rangesToPairs</span>(x, y, window, ...)</span>
<span id="cb11-4"><a href="#cb11-4"></a></span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="fu">rangesToPairs</span>(x, y, within, ...)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><br> <span style="font-size:22pt">Different ways to generate paired-ranges from range objects (i.e.&nbsp;BED files).</span></p>
</div><div class="column middle" style="width:50%;">
<p><img data-src="images/Clipboard_2022-08-22-23-17-19.png" style="width:80.0%"></p>
</div>
</div>
</section>
<section id="extractingaggregating-interactions" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">Extracting/Aggregating Interactions</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><strong>Extracting Hi-C Pixels</strong></p>
<div class="sourceCode" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy" data-id="quarto-animate-code"><code class="sourceCode r hljs"><span id="cb12-1" class="hljs-ln-code"><a href="#cb12-1"></a><span class="fu">pullHicPixels</span>(x, hic, ...)</span>
<span id="cb12-2" class="hljs-ln-code"><a href="#cb12-2"></a></span>
<span id="cb12-3" class="hljs-ln-code"><a href="#cb12-3"></a><span class="fu">pullHicMatrices</span>(x, hic, buffer, ...)</span>
<span id="cb12-4" class="hljs-ln-code"><a href="#cb12-4"></a></span>
<span id="cb12-5" class="hljs-ln-code"><a href="#cb12-5"></a><span class="fu">pullHicDomains</span>(x, hic, ...)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><span style="font-size:22pt"><strong>Pull</strong> different subsets of Hi-C data</span></p>
</div><div class="column middle" style="width:50%;">
<p><img data-src="images/apa.png" style="width:40.0%"></p>
</div>
</div>
</section>
<section id="extractingaggregating-interactions-1" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">Extracting/Aggregating Interactions</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><strong>Extracting Hi-C Pixels</strong></p>
<div class="sourceCode" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="fu">pullHicPixels</span>(x, hic, ...)</span>
<span id="cb13-2"><a href="#cb13-2"></a></span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="fu">pullHicMatrices</span>(x, hic, buffer, ...)</span>
<span id="cb13-4"><a href="#cb13-4"></a></span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="fu">pullHicDomains</span>(x, hic, ...)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Aggregating Hi-C Contacts</strong></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="fu">aggregate</span>(x, hic, shape, fun, ...)</span>
<span id="cb14-2"><a href="#cb14-2"></a></span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="fu">summarize</span>(x, hic, ...)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><span style="font-size:22pt"><strong>Pull</strong> different subsets of Hi-C data</span></p>
<p><span style="font-size:22pt"><strong>Aggregate</strong> or <strong>summarize</strong> in custom shapes</span></p>
</div><div class="column middle" style="width:50%;">
<p><img data-src="images/apa_aggregate.png" style="width:40.0%"></p>
</div>
</div>
</section></section>
<section>
<section id="workflows" class="title-slide slide level1 center middle">
<h1>Workflows</h1>

</section>
<section id="section-2" class="slide level2 middle" data-auto-animate="true">
<h2 data-id="quarto-animate-title"></h2>
<div style="height:50%; width:100%; margin:auto;">
<p><img data-src="images/diffLoopCalling.png" style="width:50.0%"></p>
</div>
<div style="height:50%; width:100%; margin:auto;">
<p>Differential loop calling</p>
</div>
<div class="footer">
<p><span class="citation" data-cites="ahn2021">Ahn et al. (<a href="#/ref-ahn2021" role="doc-biblioref" onclick="return false;">2021</a>)</span></p>
</div>
</section>
<section id="section-3" class="slide level2 middle" data-auto-animate="true">
<h2 data-id="quarto-animate-title"></h2>
<div style="height:50%; width:100%; margin:auto;">
<p><img data-src="images/apa_analysis2.png" style="width:100.0%"></p>
</div>
<div style="height:50%; width:100%; margin:auto;">
<p>Inter/Intrachromosomal pile-up analysis</p>
<p><span style="color:#787878;">Differential loop calling</span></p>
</div>
<div class="footer">
<p>In collaboration with Wang Lab, Sarah Parker, Isha Sahasrabudhe, Yoseli Quiroga and JP Flores.</p>
</div>
</section>
<section id="section-4" class="slide level2 middle" data-auto-animate="true">
<h2 data-id="quarto-animate-title"></h2>
<div style="height:50%; width:100%; margin:auto;">
<p><img data-src="images/loopDecay.png" style="width:35.0%"></p>
</div>
<div style="height:50%; width:100%; margin:auto;">
<p>Characterizing loop signal decay</p>
<p><span style="color:#787878;">Inter/Intrachromosomal pile-up analysis</span></p>
<p><span style="color:#989898;">Differential loop calling</span></p>
</div>
</section>
<section id="section-5" class="slide level2 middle" data-auto-animate="true">
<h2 data-id="quarto-animate-title"></h2>
<div style="height:50%; width:100%; margin:auto;">
<p><img data-src="images/contact_enrichment1.png" height="300"> <img data-src="images/contact_enrichment2.png" height="300"></p>
</div>
<div style="height:50%; width:100%; margin:auto;">
<p>Target contact enrichment</p>
<p><span style="color:#787878;">Characterizing loop signal decay</span></p>
<p><span style="color:#989898;">Inter/Intrachromosomal pile-up analysis</span></p>
<p><span style="color:   #B8B8B8;">Differential loop calling</span></p>
</div>
<div class="footer">
<p><span class="citation" data-cites="kelly2022">Kelly et al. (<a href="#/ref-kelly2022" role="doc-biblioref" onclick="return false;">2022</a>)</span></p>
</div>
</section>
<section id="section-6" class="slide level2 middle" data-auto-animate="true">
<h2 data-id="quarto-animate-title"></h2>
<div style="height:50%; width:100%; margin:auto;">
<p><img data-src="images/deNovoTADs.png" style="width:25.0%"></p>
</div>
<div style="height:50%; width:100%; margin:auto;">
<p>Identifying <em>de Novo</em> loops/TADs</p>
<p><span style="color:#787878;">Target contact enrichment</span></p>
<p><span style="color:#989898;">Characterizing loop signal decay</span></p>
<p><span style="color: #B8B8B8;">Inter/Intrachromosomal pile-up analysis</span></p>
<p><span style="color:   #D8D8D8;">Differential loop calling</span></p>
</div>
<div class="footer">
<p>In collaboration with Marielle Bond</p>
</div>
</section>
<section id="section-7" class="slide level2 middle" data-auto-animate="true">
<h2 data-id="quarto-animate-title"></h2>
<div style="height:50%; width:100%; margin:auto;">
<p><img data-src="images/deNovoTADs.png" style="width:25.0%"></p>
</div>
<div style="height:50%; width:100%; margin:auto;">
<p><strong>Identifying <em>de Novo</em> loops/TADs</strong></p>
<p><span style="color:#787878;">Target contact enrichment</span></p>
<p><span style="color:#989898;">Characterizing loop signal decay</span></p>
<p><span style="color: #B8B8B8;">Inter/Intrachromosomal pile-up analysis</span></p>
<p><span style="color:   #D8D8D8;">Differential loop calling</span></p>
</div>
<div class="footer">
<p>In collaboration with Marielle Bond</p>
</div>
</section></section>
<section>
<section id="example-calling-de-novo-loops" class="title-slide slide level1 center middle" data-auto-animate="true">
<h1 data-id="quarto-animate-title">Example: Calling <em>de novo</em> loops</h1>

</section>
<section id="prepare-the-data" class="slide level2">
<h2>Prepare the data</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb15" data-code-line-numbers="|5-8|10-14|15"><pre class="sourceCode numberSource r code-overflow-scroll number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="do">## Load required packages</span></span>
<span id="cb15-2"><a href="#cb15-2"></a><span class="fu">library</span>(data.table)</span>
<span id="cb15-3"><a href="#cb15-3"></a><span class="fu">library</span>(mariner)</span>
<span id="cb15-4"><a href="#cb15-4"></a></span>
<span id="cb15-5"><a href="#cb15-5"></a><span class="do">## List loop files for each timepoint</span></span>
<span id="cb15-6"><a href="#cb15-6"></a>bedpeFiles <span class="ot">&lt;-</span></span>
<span id="cb15-7"><a href="#cb15-7"></a>  <span class="fu">list.files</span>(<span class="at">path =</span> <span class="st">"data/loops"</span>,</span>
<span id="cb15-8"><a href="#cb15-8"></a>             <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb15-9"><a href="#cb15-9"></a></span>
<span id="cb15-10"><a href="#cb15-10"></a><span class="do">## Read in data and format as GInteractions</span></span>
<span id="cb15-11"><a href="#cb15-11"></a>giList <span class="ot">&lt;-</span></span>
<span id="cb15-12"><a href="#cb15-12"></a>  <span class="fu">lapply</span>(bedpeFiles, fread) <span class="sc">|&gt;</span></span>
<span id="cb15-13"><a href="#cb15-13"></a>  <span class="fu">lapply</span>(as_ginteractions) <span class="sc">|&gt;</span></span>
<span id="cb15-14"><a href="#cb15-14"></a>  <span class="fu">setNames</span>(<span class="fu">gsub</span>(<span class="st">"[^0-9]+"</span>, <span class="st">""</span>, bedpeFiles))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="prepare-the-data-1" class="slide level2">
<h2>Prepare the data</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb16" data-code-line-numbers="true"><pre class="sourceCode numberSource r code-overflow-scroll number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>giList</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`0`
GInteractions object with 10204 interactions and 9 metadata columns:
          seqnames1             ranges1     seqnames2             ranges2 |
              &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |
      [1]      chr9   89605000-89610000 ---      chr9   89670000-89675000 |
      [2]      chr9   27115000-27120000 ---      chr9   27270000-27275000 |
      [3]      chr9   33405000-33410000 ---      chr9   33610000-33615000 |
      [4]      chr9 108285000-108290000 ---      chr9 108560000-108565000 |
      [5]      chr9 128655000-128660000 ---      chr9 128690000-128695000 |
      ...       ...                 ... ...       ...                 ... .
  [10200]     chr17   35970000-35975000 ---     chr17   36480000-36485000 |
  [10201]     chr17   65275000-65280000 ---     chr17   66170000-66175000 |
  [10202]     chr17   27350000-27355000 ---     chr17   27595000-27600000 |
  [10203]     chr17   17370000-17375000 ---     chr17   17460000-17465000 |
  [10204]     chr17   17355000-17360000 ---     chr17   17640000-17645000 |
                color APScoreAvg ProbabilityofEnrichment RegAPScoreAvg
          &lt;character&gt;  &lt;numeric&gt;               &lt;numeric&gt;     &lt;numeric&gt;
      [1]       0,0,0    2.06468                0.917966       1.43249
      [2]       0,0,0    2.55654                0.983957       1.40194
      [3]       0,0,0    2.55579                0.983745       1.67442
      [4]       0,0,0    2.57163                0.943240       1.71112
      [5]       0,0,0    1.83553                0.936212       1.35813
      ...         ...        ...                     ...           ...
  [10200]       0,0,0    8.69385                1.000000       5.89909
  [10201]       0,0,0    3.48654                0.999539       2.41980
  [10202]       0,0,0    3.01999                0.992171       2.19997
  [10203]       0,0,0    2.76950                0.961080       2.16948
  [10204]       0,0,0    7.66937                0.999996       3.88862
          Avg_diffMaxNeihgboor_1 Avg_diffMaxNeihgboor_2       avg       std
                       &lt;numeric&gt;              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
      [1]               0.777091               0.807942   1.81013  0.425933
      [2]               1.881285               1.975818   2.46068  0.806084
      [3]               1.303592               1.568077   2.96105  0.677174
      [4]               0.835542               1.082364   2.12652  0.537152
      [5]               0.658496               0.789343   2.16715  0.272499
      ...                    ...                    ...       ...       ...
  [10200]               1.205008               1.883595   3.87843  0.532255
  [10201]               2.271491               3.687016   8.06904  1.210339
  [10202]               1.274382               2.298726   5.78719  0.665814
  [10203]               0.716013               0.741013   2.61012  0.336476
  [10204]               5.091658               7.040526   7.91967  1.723447
              value
          &lt;numeric&gt;
      [1]   2.50088
      [2]   4.13293
      [3]   4.11980
      [4]   2.86923
      [5]   2.75248
      ...       ...
  [10200]   4.94954
  [10201]  10.08814
  [10202]   6.91998
  [10203]   3.24658
  [10204]  12.44559
  -------
  regions: 17683 ranges and 0 metadata columns
  seqinfo: 21 sequences from an unspecified genome; no seqlengths

$`360`
GInteractions object with 10563 interactions and 9 metadata columns:
          seqnames1           ranges1     seqnames2           ranges2 |
              &lt;Rle&gt;         &lt;IRanges&gt;         &lt;Rle&gt;         &lt;IRanges&gt; |
      [1]      chr9 36105000-36110000 ---      chr9 36165000-36170000 |
      [2]      chr9 33370000-33375000 ---      chr9 33705000-33710000 |
      [3]      chr9 19500000-19505000 ---      chr9 19740000-19745000 |
      [4]      chr9 28525000-28530000 ---      chr9 32090000-32095000 |
      [5]      chr9 14460000-14465000 ---      chr9 14760000-14765000 |
      ...       ...               ... ...       ...               ... .
  [10559]     chr17 35165000-35170000 ---     chr17 35535000-35540000 |
  [10560]     chr17 46940000-46945000 ---     chr17 47000000-47005000 |
  [10561]     chr17 59315000-59320000 ---     chr17 59920000-59925000 |
  [10562]     chr17 17355000-17360000 ---     chr17 17640000-17645000 |
  [10563]     chr17 33780000-33785000 ---     chr17 33945000-33950000 |
                color APScoreAvg ProbabilityofEnrichment RegAPScoreAvg
          &lt;character&gt;  &lt;numeric&gt;               &lt;numeric&gt;     &lt;numeric&gt;
      [1]       0,0,0    1.96170                0.973668       1.43473
      [2]       0,0,0    2.55982                0.986425       1.57297
      [3]       0,0,0    2.72704                0.978201       1.69458
      [4]       0,0,0    5.50624                0.999999       3.05943
      [5]       0,0,0    3.92282                0.990343       2.62382
      ...         ...        ...                     ...           ...
  [10559]       0,0,0    3.61352                0.993416       2.55324
  [10560]       0,0,0    2.40272                0.951612       1.95457
  [10561]       0,0,0    2.47925                0.953740       1.71784
  [10562]       0,0,0    4.96387                0.999717       3.20697
  [10563]       0,0,0    3.70767                0.995413       2.31589
          Avg_diffMaxNeihgboor_1 Avg_diffMaxNeihgboor_2       avg       std
                       &lt;numeric&gt;              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
      [1]               0.777205                1.13810   2.94650  0.426781
      [2]               1.543407                1.81842   2.92802  1.010820
      [3]               1.401216                1.56290   2.58078  1.052888
      [4]               8.563278                9.44171  12.15307  3.918098
      [5]               1.287969                1.75705   3.49569  0.873397
      ...                    ...                    ...       ...       ...
  [10559]               1.475400               1.565502   3.71220  0.674091
  [10560]               0.348354               0.708525   2.71917  0.269916
  [10561]               0.978716               0.985647   2.20383  0.491411
  [10562]               2.140706               3.449179   6.26952  1.250420
  [10563]               1.644693               2.336126   3.92308  0.895122
              value
          &lt;numeric&gt;
      [1]   3.63735
      [2]   4.29994
      [3]   3.82630
      [4]  19.76487
      [5]   4.64055
      ...       ...
  [10559]   5.02366
  [10560]   3.02882
  [10561]   3.07380
  [10562]   8.17238
  [10563]   5.38503
  -------
  regions: 18678 ranges and 0 metadata columns
  seqinfo: 22 sequences from an unspecified genome; no seqlengths

$`4320`
GInteractions object with 15338 interactions and 9 metadata columns:
          seqnames1           ranges1     seqnames2             ranges2 |
              &lt;Rle&gt;         &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |
      [1]      chr9 14460000-14465000 ---      chr9   14765000-14770000 |
      [2]      chr9 75110000-75115000 ---      chr9   75175000-75180000 |
      [3]      chr9 99180000-99185000 ---      chr9 100590000-100595000 |
      [4]      chr9 96855000-96860000 ---      chr9   97390000-97395000 |
      [5]      chr9 84315000-84320000 ---      chr9   84385000-84390000 |
      ...       ...               ... ...       ...                 ... .
  [15334]     chr17 33780000-33785000 ---     chr17   34015000-34020000 |
  [15335]     chr17   2825000-2830000 ---     chr17     3450000-3455000 |
  [15336]     chr17 17355000-17360000 ---     chr17   17640000-17645000 |
  [15337]     chr17 48800000-48805000 ---     chr17   48880000-48885000 |
  [15338]     chr17 33780000-33785000 ---     chr17   33945000-33950000 |
                color APScoreAvg ProbabilityofEnrichment RegAPScoreAvg
          &lt;character&gt;  &lt;numeric&gt;               &lt;numeric&gt;     &lt;numeric&gt;
      [1]       0,0,0    4.58610                0.998150       3.34278
      [2]       0,0,0    2.31151                0.940716       1.70304
      [3]       0,0,0    4.46335                0.998697       3.03694
      [4]       0,0,0    6.01059                0.999624       3.96528
      [5]       0,0,0    2.42940                0.927548       1.60290
      ...         ...        ...                     ...           ...
  [15334]       0,0,0    2.50723                0.984298       1.62083
  [15335]       0,0,0    2.69349                0.983494       1.73560
  [15336]       0,0,0    5.51288                0.999499       3.33775
  [15337]       0,0,0    4.25762                0.990069       2.82033
  [15338]       0,0,0    3.93666                0.993367       2.45104
          Avg_diffMaxNeihgboor_1 Avg_diffMaxNeihgboor_2       avg       std
                       &lt;numeric&gt;              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
      [1]               1.401681               1.964919   5.04712  0.847931
      [2]               0.681218               0.821609   2.22017  0.382182
      [3]               1.721599               2.456924   5.11379  0.806729
      [4]               2.645426               2.870886   5.53576  1.282828
      [5]               0.817292               0.986790   1.89862  0.400903
      ...                    ...                    ...       ...       ...
  [15334]                1.44989                1.56994   2.86559  0.592246
  [15335]                1.17808                1.69171   3.05729  0.862022
  [15336]                2.13406                3.61844   5.70327  1.184972
  [15337]                1.23694                1.81455   3.51310  0.692860
  [15338]                1.73929                2.08819   3.47014  0.742591
              value
          &lt;numeric&gt;
      [1]   6.29306
      [2]   2.82570
      [3]   6.64410
      [4]   7.88725
      [5]   2.62511
      ...       ...
  [15334]   4.15438
  [15335]   4.10447
  [15336]   7.60021
  [15337]   4.61261
  [15338]   5.01617
  -------
  regions: 25431 ranges and 0 metadata columns
  seqinfo: 22 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="cluster-and-merge-pairs" class="slide level2">
<h2>Cluster and merge pairs</h2>
<div class="sourceCode" id="cb18" data-code-line-numbers="1-5|6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="do">## Cluster and merge pairs</span></span>
<span id="cb18-2"><a href="#cb18-2"></a>mgi <span class="ot">&lt;-</span> </span>
<span id="cb18-3"><a href="#cb18-3"></a>  <span class="fu">mergePairs</span>(<span class="at">x =</span> giList,</span>
<span id="cb18-4"><a href="#cb18-4"></a>             <span class="at">radius =</span> <span class="fl">10e03</span>,</span>
<span id="cb18-5"><a href="#cb18-5"></a>             <span class="at">method =</span> <span class="st">"manhattan"</span>)</span>
<span id="cb18-6"><a href="#cb18-6"></a>mgi</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="fragment">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>MergedGInteractions object with 25367 interactions and 0 metadata columns:
          seqnames1             ranges1     seqnames2             ranges2
              &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt;
      [1]      chr9   89605000-89610000 ---      chr9   89670000-89675000
      [2]      chr9   27115000-27120000 ---      chr9   27270000-27275000
      [3]      chr9   33405000-33410000 ---      chr9   33610000-33615000
      [4]      chr9 128655000-128660000 ---      chr9 128690000-128695000
      [5]      chr9 136670000-136675000 ---      chr9 136715000-136720000
      ...       ...                 ... ...       ...                 ...
  [25363]     chr17   65290000-65295000 ---     chr17   65555000-65560000
  [25364]     chr17   27657500-27662500 ---     chr17   27875000-27880000
  [25365]     chr17   29370000-29375000 ---     chr17   29530000-29535000
  [25366]     chr17   77537500-77542500 ---     chr17   78010000-78015000
  [25367]     chr17     7910000-7915000 ---     chr17     7985000-7990000
  -------
  regions: 41113 ranges and 0 metadata columns
  seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</div>
</section>
<section id="closer-look-at-the-mergedginteractions-object" class="slide level2 scrollable">
<h2>Closer look at the MergedGInteractions object</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb20" data-code-line-numbers="true"><pre class="sourceCode numberSource r code-overflow-scroll number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="do">## How did it merge?</span></span>
<span id="cb20-2"><a href="#cb20-2"></a><span class="fu">selectionMethod</span>(mgi)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Mean of modes"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb22" data-code-line-numbers="true"><pre class="sourceCode numberSource r code-overflow-scroll number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="do">## What are the clusters?</span></span>
<span id="cb22-2"><a href="#cb22-2"></a><span class="fu">tail</span>(mgi)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>MergedGInteractions object with 6 interactions and 0 metadata columns:
      seqnames1           ranges1     seqnames2           ranges2
          &lt;Rle&gt;         &lt;IRanges&gt;         &lt;Rle&gt;         &lt;IRanges&gt;
  [1]     chr17 68250000-68255000 ---     chr17 68757500-68762500
  [2]     chr17 65290000-65295000 ---     chr17 65555000-65560000
  [3]     chr17 27657500-27662500 ---     chr17 27875000-27880000
  [4]     chr17 29370000-29375000 ---     chr17 29530000-29535000
  [5]     chr17 77537500-77542500 ---     chr17 78010000-78015000
  [6]     chr17   7910000-7915000 ---     chr17   7985000-7990000
  -------
  regions: 41113 ranges and 0 metadata columns
  seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24" data-code-line-numbers="true"><pre class="sourceCode numberSource r code-overflow-scroll number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="fu">getPairClusters</span>(<span class="fu">tail</span>(mgi))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
   seqnames1   start1     end1 width1 strand1 seqnames2   start2     end2
1:     chr17 68250000 68255000   5001       *     chr17 68757500 68762500
2:     chr17 68250000 68255000   5001       *     chr17 68757500 68762500
   width2 strand2 color APScoreAvg ProbabilityofEnrichment RegAPScoreAvg
1:   5001       * 0,0,0   3.283172               0.9775382      2.183511
2:   5001       * 0,0,0   3.766162               0.9794906      2.477963
   Avg_diffMaxNeihgboor_1 Avg_diffMaxNeihgboor_2      avg       std    value
1:               1.267313               1.353121 2.669831 0.6812279 3.796332
2:               1.154502               1.500289 2.861052 0.4824793 3.887276
    src
1:  360
2: 4320

[[2]]
   seqnames1   start1     end1 width1 strand1 seqnames2   start2     end2
1:     chr17 65290000 65295000   5001       *     chr17 65555000 65560000
2:     chr17 65290000 65295000   5001       *     chr17 65555000 65560000
   width2 strand2 color APScoreAvg ProbabilityofEnrichment RegAPScoreAvg
1:   5001       * 0,0,0   2.301621               0.9782407      1.811257
2:   5001       * 0,0,0   4.184173               0.9985222      2.732198
   Avg_diffMaxNeihgboor_1 Avg_diffMaxNeihgboor_2      avg       std    value
1:              0.5883462              0.9801798 3.305139 0.3782773 3.828114
2:              1.7415318              2.6633184 4.969868 0.9183704 6.517896
    src
1:  360
2: 4320

[[3]]
   seqnames1   start1     end1 width1 strand1 seqnames2   start2     end2
1:     chr17 27657500 27662500   5001       *     chr17 27875000 27880000
2:     chr17 27657500 27662500   5001       *     chr17 27875000 27880000
   width2 strand2 color APScoreAvg ProbabilityofEnrichment RegAPScoreAvg
1:   5001       * 0,0,0   3.730618                0.978591      2.377280
2:   5001       * 0,0,0   4.236078                1.000000      2.490079
   Avg_diffMaxNeihgboor_1 Avg_diffMaxNeihgboor_2      avg       std    value
1:               1.079278               1.639414 2.884983 0.5946189 3.844341
2:               1.224047               1.691289 2.488413 0.6791529 3.576455
    src
1:  360
2: 4320

[[4]]
   seqnames1   start1     end1 width1 strand1 seqnames2   start2     end2
1:     chr17 29370000 29375000   5001       *     chr17 29530000 29535000
2:     chr17 29370000 29375000   5001       *     chr17 29530000 29535000
   width2 strand2 color APScoreAvg ProbabilityofEnrichment RegAPScoreAvg
1:   5001       * 0,0,0   2.526620               0.9829903      1.893225
2:   5001       * 0,0,0   2.970785               0.9883910      2.120753
   Avg_diffMaxNeihgboor_1 Avg_diffMaxNeihgboor_2      avg       std    value
1:              0.8547273               1.168578 3.314633 0.3949473 4.074391
2:              1.0702407               1.457258 3.505113 0.5895870 4.456438
    src
1:  360
2: 4320

[[5]]
   seqnames1   start1     end1 width1 strand1 seqnames2   start2     end2
1:     chr17 77537500 77542500   5001       *     chr17 78010000 78015000
2:     chr17 77537500 77542500   5001       *     chr17 78010000 78015000
   width2 strand2 color APScoreAvg ProbabilityofEnrichment RegAPScoreAvg
1:   5001       * 0,0,0   3.200017               0.9959567      2.346039
2:   5001       * 0,0,0   3.623490               0.9977793      2.309720
   Avg_diffMaxNeihgboor_1 Avg_diffMaxNeihgboor_2      avg       std    value
1:               1.401516               1.597327 4.265478 0.7819768 5.511270
2:               1.775418               2.574010 4.532385 0.7181578 6.110535
    src
1:  360
2: 4320

[[6]]
   seqnames1  start1    end1 width1 strand1 seqnames2  start2    end2 width2
1:     chr17 7910000 7915000   5001       *     chr17 7985000 7990000   5001
2:     chr17 7910000 7915000   5001       *     chr17 7985000 7990000   5001
   strand2 color APScoreAvg ProbabilityofEnrichment RegAPScoreAvg
1:       * 0,0,0   2.404220               0.9907429      1.626797
2:       * 0,0,0   2.427026               0.9866196      1.642203
   Avg_diffMaxNeihgboor_1 Avg_diffMaxNeihgboor_2      avg       std    value
1:               1.340799               1.695594 3.491029 0.6438525 4.682849
2:               1.045785               1.657023 3.384829 0.4991671 4.314417
    src
1:  360
2: 4320</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26" data-code-line-numbers="true"><pre class="sourceCode numberSource r code-overflow-scroll number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="do">## Let's recover TAD names</span></span>
<span id="cb26-2"><a href="#cb26-2"></a>mgi <span class="ot">&lt;-</span> <span class="fu">aggPairMcols</span>(mgi, <span class="at">columns =</span> <span class="st">"APScoreAvg"</span>, <span class="at">funs =</span> <span class="st">"mean"</span>)</span>
<span id="cb26-3"><a href="#cb26-3"></a><span class="fu">tail</span>(mgi)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>MergedGInteractions object with 6 interactions and 1 metadata column:
      seqnames1           ranges1     seqnames2           ranges2 |
          &lt;Rle&gt;         &lt;IRanges&gt;         &lt;Rle&gt;         &lt;IRanges&gt; |
  [1]     chr17 68250000-68255000 ---     chr17 68757500-68762500 |
  [2]     chr17 65290000-65295000 ---     chr17 65555000-65560000 |
  [3]     chr17 27657500-27662500 ---     chr17 27875000-27880000 |
  [4]     chr17 29370000-29375000 ---     chr17 29530000-29535000 |
  [5]     chr17 77537500-77542500 ---     chr17 78010000-78015000 |
  [6]     chr17   7910000-7915000 ---     chr17   7985000-7990000 |
      mean.APScoreAvg
            &lt;numeric&gt;
  [1]         3.52467
  [2]         3.24290
  [3]         3.98335
  [4]         2.74870
  [5]         3.41175
  [6]         2.41562
  -------
  regions: 41113 ranges and 0 metadata columns
  seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="calling-de-novo-with-subsetbysource" class="slide level2 scrollable">
<h2>Calling <em>de novo</em> with <code>subsetBySource()</code></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb28" data-code-line-numbers="true"><pre class="sourceCode numberSource r code-overflow-scroll number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a><span class="do">## Available sources</span></span>
<span id="cb28-2"><a href="#cb28-2"></a><span class="fu">sources</span>(mgi)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "0"    "360"  "4320"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30" data-code-line-numbers="true"><pre class="sourceCode numberSource r code-overflow-scroll number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="do">## Subsetting loops by source</span></span>
<span id="cb30-2"><a href="#cb30-2"></a><span class="fu">subsetBySource</span>(mgi)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`0`
MergedGInteractions object with 4425 interactions and 1 metadata column:
         seqnames1             ranges1     seqnames2             ranges2 |
             &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |
     [1]      chr9   89605000-89610000 ---      chr9   89670000-89675000 |
     [2]      chr9   27115000-27120000 ---      chr9   27270000-27275000 |
     [3]      chr9   33405000-33410000 ---      chr9   33610000-33615000 |
     [4]      chr9 128655000-128660000 ---      chr9 128690000-128695000 |
     [5]      chr9 136670000-136675000 ---      chr9 136715000-136720000 |
     ...       ...                 ... ...       ...                 ... .
  [4421]     chr17   28970000-28975000 ---     chr17   29215000-29220000 |
  [4422]     chr17   37525000-37530000 ---     chr17   37785000-37790000 |
  [4423]     chr17   36490000-36495000 ---     chr17   36525000-36530000 |
  [4424]     chr17   35970000-35975000 ---     chr17   36480000-36485000 |
  [4425]     chr17   17370000-17375000 ---     chr17   17460000-17465000 |
         mean.APScoreAvg
               &lt;numeric&gt;
     [1]         2.06468
     [2]         2.55654
     [3]         2.55579
     [4]         1.83553
     [5]         1.85282
     ...             ...
  [4421]         2.84342
  [4422]         2.49312
  [4423]         3.05257
  [4424]         8.69385
  [4425]         2.76950
  -------
  regions: 41113 ranges and 0 metadata columns
  seqinfo: 23 sequences from an unspecified genome; no seqlengths

$`360`
MergedGInteractions object with 4635 interactions and 1 metadata column:
         seqnames1             ranges1     seqnames2             ranges2 |
             &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |
     [1]      chr9   33370000-33375000 ---      chr9   33705000-33710000 |
     [2]      chr9   19500000-19505000 ---      chr9   19740000-19745000 |
     [3]      chr9   28525000-28530000 ---      chr9   32090000-32095000 |
     [4]      chr9 114405000-114410000 ---      chr9 114915000-114920000 |
     [5]      chr9     6060000-6065000 ---      chr9     6575000-6580000 |
     ...       ...                 ... ...       ...                 ... .
  [4631]     chr17     7270000-7275000 ---     chr17     7345000-7350000 |
  [4632]     chr17   18135000-18140000 ---     chr17   18355000-18360000 |
  [4633]     chr17   38590000-38595000 ---     chr17   38655000-38660000 |
  [4634]     chr17   46940000-46945000 ---     chr17   47000000-47005000 |
  [4635]     chr17   59315000-59320000 ---     chr17   59920000-59925000 |
         mean.APScoreAvg
               &lt;numeric&gt;
     [1]         2.55982
     [2]         2.72704
     [3]         5.50624
     [4]         2.74619
     [5]         2.69432
     ...             ...
  [4631]         2.47831
  [4632]         4.12110
  [4633]         2.03401
  [4634]         2.40272
  [4635]         2.47925
  -------
  regions: 41113 ranges and 0 metadata columns
  seqinfo: 23 sequences from an unspecified genome; no seqlengths

$`4320`
MergedGInteractions object with 8698 interactions and 1 metadata column:
         seqnames1             ranges1     seqnames2             ranges2 |
             &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |
     [1]      chr9   75110000-75115000 ---      chr9   75175000-75180000 |
     [2]      chr9   99180000-99185000 ---      chr9 100590000-100595000 |
     [3]      chr9   84315000-84320000 ---      chr9   84385000-84390000 |
     [4]      chr9 116300000-116305000 ---      chr9 116485000-116490000 |
     [5]      chr9 114910000-114915000 ---      chr9 115010000-115015000 |
     ...       ...                 ... ...       ...                 ... .
  [8694]     chr17   68915000-68920000 ---     chr17   69230000-69235000 |
  [8695]     chr17   30565000-30570000 ---     chr17   30700000-30705000 |
  [8696]     chr17   29585000-29590000 ---     chr17   30085000-30090000 |
  [8697]     chr17   33780000-33785000 ---     chr17   34015000-34020000 |
  [8698]     chr17     2825000-2830000 ---     chr17     3450000-3455000 |
         mean.APScoreAvg
               &lt;numeric&gt;
     [1]         2.31151
     [2]         4.46335
     [3]         2.42940
     [4]         2.11791
     [5]         4.01746
     ...             ...
  [8694]         2.14816
  [8695]         2.63569
  [8696]         3.84359
  [8697]         2.50723
  [8698]         2.69349
  -------
  regions: 41113 ranges and 0 metadata columns
  seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb32" data-code-line-numbers="true"><pre class="sourceCode numberSource r code-overflow-scroll number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a><span class="do">## Transient loops</span></span>
<span id="cb32-2"><a href="#cb32-2"></a><span class="fu">subsetBySource</span>(<span class="at">x =</span> mgi,</span>
<span id="cb32-3"><a href="#cb32-3"></a>               <span class="at">include =</span> <span class="st">"360"</span>,</span>
<span id="cb32-4"><a href="#cb32-4"></a>               <span class="at">exclude =</span> <span class="fu">c</span>(<span class="st">"0"</span>, <span class="st">"4320"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>MergedGInteractions object with 4635 interactions and 1 metadata column:
         seqnames1             ranges1     seqnames2             ranges2 |
             &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |
     [1]      chr9   33370000-33375000 ---      chr9   33705000-33710000 |
     [2]      chr9   19500000-19505000 ---      chr9   19740000-19745000 |
     [3]      chr9   28525000-28530000 ---      chr9   32090000-32095000 |
     [4]      chr9 114405000-114410000 ---      chr9 114915000-114920000 |
     [5]      chr9     6060000-6065000 ---      chr9     6575000-6580000 |
     ...       ...                 ... ...       ...                 ... .
  [4631]     chr17     7270000-7275000 ---     chr17     7345000-7350000 |
  [4632]     chr17   18135000-18140000 ---     chr17   18355000-18360000 |
  [4633]     chr17   38590000-38595000 ---     chr17   38655000-38660000 |
  [4634]     chr17   46940000-46945000 ---     chr17   47000000-47005000 |
  [4635]     chr17   59315000-59320000 ---     chr17   59920000-59925000 |
         mean.APScoreAvg
               &lt;numeric&gt;
     [1]         2.55982
     [2]         2.72704
     [3]         5.50624
     [4]         2.74619
     [5]         2.69432
     ...             ...
  [4631]         2.47831
  [4632]         4.12110
  [4633]         2.03401
  [4634]         2.40272
  [4635]         2.47925
  -------
  regions: 41113 ranges and 0 metadata columns
  seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb34" data-code-line-numbers="true"><pre class="sourceCode numberSource r code-overflow-scroll number-lines code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a><span class="do">## De novo (must be in 72 but not in 0)</span></span>
<span id="cb34-2"><a href="#cb34-2"></a>deNovo <span class="ot">&lt;-</span></span>
<span id="cb34-3"><a href="#cb34-3"></a>  <span class="fu">subsetBySource</span>(<span class="at">x =</span> mgi,</span>
<span id="cb34-4"><a href="#cb34-4"></a>                 <span class="at">include =</span> <span class="st">"4320"</span>,</span>
<span id="cb34-5"><a href="#cb34-5"></a>                 <span class="at">exclude =</span> <span class="st">"0"</span>)</span>
<span id="cb34-6"><a href="#cb34-6"></a>deNovo</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>MergedGInteractions object with 10528 interactions and 1 metadata column:
          seqnames1             ranges1     seqnames2             ranges2 |
              &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |
      [1]      chr9   75110000-75115000 ---      chr9   75175000-75180000 |
      [2]      chr9   99180000-99185000 ---      chr9 100590000-100595000 |
      [3]      chr9   84315000-84320000 ---      chr9   84385000-84390000 |
      [4]      chr9 116300000-116305000 ---      chr9 116485000-116490000 |
      [5]      chr9 114910000-114915000 ---      chr9 115010000-115015000 |
      ...       ...                 ... ...       ...                 ... .
  [10524]     chr17   65290000-65295000 ---     chr17   65555000-65560000 |
  [10525]     chr17   27657500-27662500 ---     chr17   27875000-27880000 |
  [10526]     chr17   29370000-29375000 ---     chr17   29530000-29535000 |
  [10527]     chr17   77537500-77542500 ---     chr17   78010000-78015000 |
  [10528]     chr17     7910000-7915000 ---     chr17     7985000-7990000 |
          mean.APScoreAvg
                &lt;numeric&gt;
      [1]         2.31151
      [2]         4.46335
      [3]         2.42940
      [4]         2.11791
      [5]         4.01746
      ...             ...
  [10524]         3.24290
  [10525]         3.98335
  [10526]         2.74870
  [10527]         3.41175
  [10528]         2.41562
  -------
  regions: 41113 ranges and 0 metadata columns
  seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="format-for-visualization" class="slide level2">
<h2>Format for visualization</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb36" data-code-line-numbers="true"><pre class="sourceCode numberSource r code-overflow-scroll number-lines code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a><span class="do">## Assign to Hi-C bins</span></span>
<span id="cb36-2"><a href="#cb36-2"></a>deNovo <span class="ot">&lt;-</span> </span>
<span id="cb36-3"><a href="#cb36-3"></a>  <span class="fu">binPairs</span>(<span class="at">x =</span> deNovo,</span>
<span id="cb36-4"><a href="#cb36-4"></a>           <span class="at">binSize =</span> <span class="fl">10e03</span>)</span>
<span id="cb36-5"><a href="#cb36-5"></a></span>
<span id="cb36-6"><a href="#cb36-6"></a><span class="do">## Change chromosome style</span></span>
<span id="cb36-7"><a href="#cb36-7"></a><span class="fu">library</span>(<span class="at">package =</span> <span class="st">"GenomeInfoDb"</span>,</span>
<span id="cb36-8"><a href="#cb36-8"></a>        <span class="at">include.only =</span> <span class="st">"seqlevelsStyle"</span>)</span>
<span id="cb36-9"><a href="#cb36-9"></a><span class="fu">seqlevelsStyle</span>(deNovo) <span class="ot">&lt;-</span> <span class="st">"ENSEMBL"</span></span>
<span id="cb36-10"><a href="#cb36-10"></a>deNovo</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>MergedGInteractions object with 10528 interactions and 1 metadata column:
          seqnames1             ranges1     seqnames2             ranges2 |
              &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |
      [1]         9   75110000-75120000 ---         9   75170000-75180000 |
      [2]         9   99180000-99190000 ---         9 100590000-100600000 |
      [3]         9   84310000-84320000 ---         9   84380000-84390000 |
      [4]         9 116300000-116310000 ---         9 116480000-116490000 |
      [5]         9 114910000-114920000 ---         9 115010000-115020000 |
      ...       ...                 ... ...       ...                 ... .
  [10524]        17   65290000-65300000 ---        17   65550000-65560000 |
  [10525]        17   27660000-27670000 ---        17   27870000-27880000 |
  [10526]        17   29370000-29380000 ---        17   29530000-29540000 |
  [10527]        17   77540000-77550000 ---        17   78010000-78020000 |
  [10528]        17     7910000-7920000 ---        17     7980000-7990000 |
          mean.APScoreAvg
                &lt;numeric&gt;
      [1]         2.31151
      [2]         4.46335
      [3]         2.42940
      [4]         2.11791
      [5]         4.01746
      ...             ...
  [10524]         3.24290
  [10525]         3.98335
  [10526]         2.74870
  [10527]         3.41175
  [10528]         2.41562
  -------
  regions: 17128 ranges and 0 metadata columns
  seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="visualize-with-plotgardener" class="slide level2">
<h2>Visualize with <code>plotgardener</code></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb38" data-code-line-numbers="true"><pre class="sourceCode numberSource r code-overflow-scroll number-lines code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a><span class="do">## Load libraries</span></span>
<span id="cb38-2"><a href="#cb38-2"></a><span class="fu">library</span>(plotgardener)</span>
<span id="cb38-3"><a href="#cb38-3"></a></span>
<span id="cb38-4"><a href="#cb38-4"></a><span class="do">## Define common params</span></span>
<span id="cb38-5"><a href="#cb38-5"></a>p <span class="ot">&lt;-</span> <span class="fu">pgParams</span>(<span class="at">assembly =</span> <span class="st">"hg38"</span>,</span>
<span id="cb38-6"><a href="#cb38-6"></a>              </span>
<span id="cb38-7"><a href="#cb38-7"></a>             <span class="do">## Define region of interest</span></span>
<span id="cb38-8"><a href="#cb38-8"></a>             <span class="at">chrom =</span> <span class="st">"7"</span>,</span>
<span id="cb38-9"><a href="#cb38-9"></a>             <span class="at">chromstart =</span> <span class="dv">54660000</span>,</span>
<span id="cb38-10"><a href="#cb38-10"></a>             <span class="at">chromend =</span> <span class="dv">55350000</span>,</span>
<span id="cb38-11"><a href="#cb38-11"></a>             <span class="at">resolution =</span> <span class="fl">10e03</span>,</span>
<span id="cb38-12"><a href="#cb38-12"></a>             <span class="at">zrange =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">275</span>),</span>
<span id="cb38-13"><a href="#cb38-13"></a>             <span class="at">norm =</span> <span class="st">"SCALE"</span>,</span>
<span id="cb38-14"><a href="#cb38-14"></a>             </span>
<span id="cb38-15"><a href="#cb38-15"></a>             <span class="do">## Define reference point</span></span>
<span id="cb38-16"><a href="#cb38-16"></a>             <span class="at">x =</span> <span class="fl">0.5</span>,</span>
<span id="cb38-17"><a href="#cb38-17"></a>             <span class="at">y =</span> <span class="fl">0.5</span>,</span>
<span id="cb38-18"><a href="#cb38-18"></a>             <span class="at">width =</span> <span class="dv">3</span>,</span>
<span id="cb38-19"><a href="#cb38-19"></a>             <span class="at">height =</span> <span class="dv">3</span>,</span>
<span id="cb38-20"><a href="#cb38-20"></a>             <span class="at">length =</span> <span class="dv">3</span>,</span>
<span id="cb38-21"><a href="#cb38-21"></a>             <span class="at">space =</span> <span class="fl">0.1</span>)</span>
<span id="cb38-22"><a href="#cb38-22"></a></span>
<span id="cb38-23"><a href="#cb38-23"></a><span class="do">## Initialize page</span></span>
<span id="cb38-24"><a href="#cb38-24"></a><span class="fu">pageCreate</span>(<span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="dv">4</span>, <span class="at">showGuides =</span> <span class="cn">FALSE</span>)</span>
<span id="cb38-25"><a href="#cb38-25"></a></span>
<span id="cb38-26"><a href="#cb38-26"></a><span class="do">## Hi-C plots</span></span>
<span id="cb38-27"><a href="#cb38-27"></a>top <span class="ot">&lt;-</span> </span>
<span id="cb38-28"><a href="#cb38-28"></a>  <span class="fu">plotHicSquare</span>(<span class="at">data =</span> <span class="st">"data/hic/MEGA_K562_WT_0_inter.hic"</span>,</span>
<span id="cb38-29"><a href="#cb38-29"></a>                <span class="at">params =</span> p,</span>
<span id="cb38-30"><a href="#cb38-30"></a>                <span class="at">half =</span> <span class="st">"top"</span>)</span>
<span id="cb38-31"><a href="#cb38-31"></a>bottom <span class="ot">&lt;-</span> </span>
<span id="cb38-32"><a href="#cb38-32"></a>  <span class="fu">plotHicSquare</span>(<span class="at">data =</span> <span class="st">"data/hic/MEGA_K562_WT_4320_inter.hic"</span>,</span>
<span id="cb38-33"><a href="#cb38-33"></a>                <span class="at">params =</span> p,</span>
<span id="cb38-34"><a href="#cb38-34"></a>                <span class="at">half =</span> <span class="st">"bottom"</span>)</span>
<span id="cb38-35"><a href="#cb38-35"></a></span>
<span id="cb38-36"><a href="#cb38-36"></a><span class="do">## Annotate loop pixels</span></span>
<span id="cb38-37"><a href="#cb38-37"></a><span class="fu">annoPixels</span>(<span class="at">plot =</span> bottom,</span>
<span id="cb38-38"><a href="#cb38-38"></a>           <span class="at">data =</span> deNovo,</span>
<span id="cb38-39"><a href="#cb38-39"></a>           <span class="at">shift =</span> <span class="dv">3</span>,</span>
<span id="cb38-40"><a href="#cb38-40"></a>           <span class="at">col =</span> <span class="st">'red'</span>,</span>
<span id="cb38-41"><a href="#cb38-41"></a>           <span class="at">params =</span> p)</span>
<span id="cb38-42"><a href="#cb38-42"></a></span>
<span id="cb38-43"><a href="#cb38-43"></a><span class="do">## Annotate heatmap legend</span></span>
<span id="cb38-44"><a href="#cb38-44"></a><span class="fu">annoHeatmapLegend</span>(<span class="at">plot =</span> bottom,</span>
<span id="cb38-45"><a href="#cb38-45"></a>                  <span class="at">params =</span> p,</span>
<span id="cb38-46"><a href="#cb38-46"></a>                  <span class="at">orientation =</span> <span class="st">"v"</span>,</span>
<span id="cb38-47"><a href="#cb38-47"></a>                  <span class="at">fontcolor =</span> <span class="st">"black"</span>,</span>
<span id="cb38-48"><a href="#cb38-48"></a>                  <span class="at">x =</span> p<span class="sc">$</span>x <span class="sc">+</span> p<span class="sc">$</span>width <span class="sc">+</span> p<span class="sc">$</span>space,</span>
<span id="cb38-49"><a href="#cb38-49"></a>                  <span class="at">width =</span> p<span class="sc">$</span>space,</span>
<span id="cb38-50"><a href="#cb38-50"></a>                  <span class="at">height =</span> p<span class="sc">$</span>height<span class="sc">*</span><span class="fl">0.5</span>)</span>
<span id="cb38-51"><a href="#cb38-51"></a>    </span>
<span id="cb38-52"><a href="#cb38-52"></a><span class="do">## Label genome</span></span>
<span id="cb38-53"><a href="#cb38-53"></a><span class="fu">annoGenomeLabel</span>(<span class="at">plot =</span> bottom,</span>
<span id="cb38-54"><a href="#cb38-54"></a>                <span class="at">params =</span> p,</span>
<span id="cb38-55"><a href="#cb38-55"></a>                <span class="at">y =</span> p<span class="sc">$</span>y <span class="sc">+</span> p<span class="sc">$</span>height,</span>
<span id="cb38-56"><a href="#cb38-56"></a>                <span class="at">scale =</span> <span class="st">"bp"</span>)</span>
<span id="cb38-57"><a href="#cb38-57"></a></span>
<span id="cb38-58"><a href="#cb38-58"></a><span class="do">## Text annotations</span></span>
<span id="cb38-59"><a href="#cb38-59"></a><span class="fu">plotText</span>(<span class="at">label =</span> <span class="fu">c</span>(<span class="st">"0 hrs"</span>, <span class="st">"72 hrs"</span>),</span>
<span id="cb38-60"><a href="#cb38-60"></a>         <span class="at">fontcolor =</span> <span class="st">"black"</span>,</span>
<span id="cb38-61"><a href="#cb38-61"></a>         <span class="at">x =</span> p<span class="sc">$</span>x <span class="sc">+</span> p<span class="sc">$</span>space,</span>
<span id="cb38-62"><a href="#cb38-62"></a>         <span class="at">y =</span> p<span class="sc">$</span>y <span class="sc">+</span> p<span class="sc">$</span>space,</span>
<span id="cb38-63"><a href="#cb38-63"></a>         <span class="at">just =</span> <span class="fu">c</span>(<span class="st">'left'</span>, <span class="st">'top'</span>))</span>
<span id="cb38-64"><a href="#cb38-64"></a></span>
<span id="cb38-65"><a href="#cb38-65"></a><span class="fu">plotText</span>(<span class="at">label =</span> <span class="st">"72 hrs"</span>,</span>
<span id="cb38-66"><a href="#cb38-66"></a>         <span class="at">fontcolor =</span> <span class="st">"black"</span>,</span>
<span id="cb38-67"><a href="#cb38-67"></a>         <span class="at">x =</span> p<span class="sc">$</span>x <span class="sc">+</span> p<span class="sc">$</span>width <span class="sc">-</span> p<span class="sc">$</span>space,</span>
<span id="cb38-68"><a href="#cb38-68"></a>         <span class="at">y =</span> p<span class="sc">$</span>y <span class="sc">+</span> p<span class="sc">$</span>height <span class="sc">-</span> p<span class="sc">$</span>space,</span>
<span id="cb38-69"><a href="#cb38-69"></a>         <span class="at">just =</span> <span class="fu">c</span>(<span class="st">"right"</span>, <span class="st">"bottom"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

</section><section id="visualize-with-plotgardener-output" class="slide level2 output-location-slide"><h2>Visualize with <code>plotgardener</code></h2><div class="cell output-location-slide">
<div class="cell-output-display">
<p><img data-src="mariner_files/figure-revealjs/unnamed-chunk-10-1.png" width="960"></p>
</div>
</div></section></section>
<section>
<section id="references" class="title-slide slide level1 center middle">
<h1>References</h1>

</section>
<section id="references-1" class="slide level2 smaller scrollable">
<h2>References</h2>
<div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-ahn2021" class="csl-entry" role="doc-biblioentry">
Ahn, Jeong Hyun, Eric S. Davis, Timothy A. Daugird, Shuai Zhao, Ivana Yoseli Quiroga, Hidetaka Uryu, Jie Li, et al. 2021. <span>“Phase Separation Drives Aberrant Chromatin Looping and Cancer Development.”</span> <em>Nature</em> 595 (7868): 591–95. <a href="https://doi.org/10.1038/s41586-021-03662-5">https://doi.org/10.1038/s41586-021-03662-5</a>.
</div>
<div id="ref-kelly2022" class="csl-entry" role="doc-biblioentry">
Kelly, Michael R., Kamila Wisniewska, Matthew J. Regner, Michael W. Lewis, Andrea A. Perreault, Eric S. Davis, Douglas H. Phanstiel, Joel S. Parker, and Hector L. Franco. 2022. <span>“A Multi-Omic Dissection of Super-Enhancer Driven Oncogenic Gene Expression Programs in Ovarian Cancer.”</span> <em>Nature Communications</em> 13 (1). <a href="https://doi.org/10.1038/s41467-022-31919-8">https://doi.org/10.1038/s41467-022-31919-8</a>.
</div>
</div>
</section>
<section id="acknowledgements" class="slide level2">
<h2>Acknowledgements</h2>
<p><img data-src="images/acknowledgements.png"></p>
<p><img src="images/bcb_logo.png" class="slide-logo"></p>
<div class="footer footer-default">
<p><a href="https://github.com/EricSDavis/mariner" class="uri">https://github.com/EricSDavis/mariner</a></p>
</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="mariner_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="mariner_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="mariner_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="mariner_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="mariner_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="mariner_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="mariner_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="mariner_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="mariner_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="mariner_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="mariner_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        function fireSlideChanged(previousSlide, currentSlide) {

          // dispatch for htmlwidgets
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for reveal
        if (window.Reveal) {
          window.Reveal.addEventListener("slidechanged", function(event) {
            fireSlideChanged(event.previousSlide, event.currentSlide);
          });
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        target: function(trigger) {
          return trigger.previousElementSibling;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        setTimeout(function() {
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'quarto-reveal',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          let href = ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const cites = ref.parentNode.getAttribute('data-cites').split(' ');
        tippyHover(ref, function() {
          var popup = window.document.createElement('div');
          cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    });
    </script>
    

</body></html>